pkg/actuators/machine/reconciler.go

```diff
@@ -303,3 +314,124 @@ func (r *Reconciler) getMachineInstance() (*models.PVMInstance, error) {
+		// If the networkAddress length is more than 1 means, either NodeInternalIP or NodeExternalIP is updated so return
+		r.machineScope.machine.Status.Addresses = networkAddresses
+		return nil
+	}
+	// In this case there is no IP found under instance.Networks, So try to fetch the IP from cache or DHCP server
+	// Look for DHCP IP from the cache
+	ip, ok, err := r.machineScope.dhcpIPCacheStore.GetByKey(*instance.ServerName)
+	if err != nil {
+		klog.Errorf("failed to fetch the dhcp ip fro vm : %s from cache store error: %v", *instance.ServerName, err)
+	}
+	if ok {
+		klog.Infof("found IP: %s for VM: %s from dhcp cache", ip.(vmIP).ip, *instance.ServerName)
+		networkAddresses = append(networkAddresses, corev1.NodeAddress{
+			Type:    corev1.NodeInternalIP,
+			Address: ip.(vmIP).ip,
+		})
+		r.machineScope.machine.Status.Addresses = networkAddresses
+		return nil
+	}
+	// Fetch the VM network ID
+	networkID, err := getNetworkID(r.providerSpec.Network, r.powerVSClient)
+	if err != nil {
+		errStr := fmt.Errorf("failed to fetch network id from network resource for VM: %s error: %v ", r.machine.Name, err)
+		klog.Errorf(errStr.Error())
+		return errStr
+	}
+	// Fetch the details of the network attached to the VM
+	var pvmNetwork *models.PVMInstanceNetwork
+	for _, network := range instance.Networks {
+		if network.NetworkID == *networkID {
+			pvmNetwork = network
+			klog.Infof("found network with ID %s attached to VM %s", network.NetworkID, *instance.ServerName)
+		}
+	}
+	if pvmNetwork == nil {
+		errStr := fmt.Errorf("failed to get network attached to vm %s with id %s ", *instance.ServerName, *networkID)
+		klog.Errorf(errStr.Error())
+		return errStr
+	}
+	// Get all the DHCP servers
+	dhcpServer, err := r.powerVSClient.GetDHCPServers()
+	if err != nil {
+		errStr := fmt.Errorf("failed to get DHCP server error: %v ", err)
+		klog.Errorf(errStr.Error())
+		return errStr
+	}
+	// Get the Details of DHCP server associated with the network
+	var dhcpServerDetails *models.DHCPServerDetail
+	for _, server := range dhcpServer {
+		if *server.Network.ID == *networkID {
+			klog.Infof("found DHCP server with ID %s for network ID %s", *server.Network.ID, *networkID)
+			dhcpServerDetails, err = r.powerVSClient.GetDHCPServerByID(*server.ID)
+			if err != nil || dhcpServerDetails == nil {
+				errStr := fmt.Errorf("failed to get DHCP server details for with DHCP server ID: %s error: %v ", *server.Network.ID, err)
+				klog.Errorf(errStr.Error())
+				return errStr
+			}
+		}
+	}
+	// Fetch the VM IP using VM's mac from DHCP server lease
+	var internalIP *string
+	for _, lease := range dhcpServerDetails.Leases {
+		if *lease.InstanceMacAddress == pvmNetwork.MacAddress {
+			klog.Infof("found internal ip %s for VM %s from dhcp lease", *lease.InstanceIP, *instance.ServerName)
+			internalIP = lease.InstanceIP
+		}
+	}
+	if internalIP == nil {
+		errStr := fmt.Errorf("failed to get internal IP for VM %s ", *instance.ServerName)
```

Comment:
can this be more meaningful message like lease not found for the so and so server in so and so dhcp service